platform-sdk/swirlds-common/src/main/java/com/swirlds/common/wiring/counters/BackpressureObjectCounter.java

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright (C) 2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.common.wiring.counters;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.time.Duration;
+import java.util.Objects;
+import java.util.concurrent.ForkJoinPool;
+import java.util.concurrent.ForkJoinPool.ManagedBlocker;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.LongUnaryOperator;
+
+/**
+ * A utility for counting the number of objects in various parts of the pipeline. Will apply backpressure if the number
+ * of objects exceeds a specified capacity.
+ */
+public class BackpressureObjectCounter extends ObjectCounter {
+
+    private final String name;
+    private final AtomicLong count = new AtomicLong(0);
+    private final LongUnaryOperator increment;
+
+    /**
+     * When back pressure needs to be applied due to lack of capacity, this object is used to efficiently sleep on the
+     * fork join pool.
+     */
+    private final ManagedBlocker onRampBlocker;
+
+    /**
+     * When waiting for the count to reach zero, this object is used to efficiently sleep on the fork join pool.
+     */
+    private final ManagedBlocker waitUntilEmptyBlocker;
+
+    /**
+     * Constructor.
+     *
+     * @param name          the name of the object counter, used creating more informative exceptions
+     * @param capacity      the maximum number of objects that can be in the part of the system that this object is
+     *                      being used to monitor before backpressure is applied
+     * @param sleepDuration when a method needs to block, the duration to sleep while blocking
+     */
+    public BackpressureObjectCounter(
+            @NonNull final String name, final long capacity, @NonNull final Duration sleepDuration) {
+        if (capacity <= 0) {
+            throw new IllegalArgumentException("Capacity must be greater than zero");
+        }
+
+        this.name = Objects.requireNonNull(name);
+
+        increment = count -> {
+            if (count >= capacity) {
+                throw new NoCapacityException();
```

Comment:
There is no benefit in using an exception here and a lot of cost (both heap/GC and CPU).
It ends up in a busy wait loop anyway (`ForkJoinPool.managedBlock()` -> `onRampBlocker.isReleasable()` -> `count.compareAndSet()`), so, why add costly complexity that will be on every radar screen for real code.
The ideal situation is when the back-pressure mechanism is not invoked (all covered consumers are faster then the producer) but it should not exacerbate the condition, by creating pressure on the GC for one thing, when that's not the case.

Reply:
I've changed my implementation as suggested.